### ISSUE TITLE
Bug: NaN/infinity problem - understand when/if non-finite values are legit

### ISSUE THREAD

**Issue by poja** [poja/Cattus#119]
This is a pretty weird problem and we don't have a good indication that the library is working right.
We should better understand what's up with these NaNs.

I found these:
* NaN on cpu but not on GPU https://github.com/tensorflow/tensorflow/issues/48205
* Someone talking about the general issue of non-finite values in tensorflow. Gives some tools for debugging. https://stable-baselines.readthedocs.io/en/master/guide/checking_nan.html

Some possible approaches: 
* Train until we get NaNs and then check the weights
* Train until we get NaNs and then see what the different results are when computing multiple times
* Isolate the problem - reduce the network complexity until there are no NaNs and see if there was something specific that caused it or it just starts happening when the network is big enough.

**Comment by barakugav** [comment]
We should check if the problem is reproducible when running from Python. The problem might be in our network structure.
Did you try changing the batch norm axis?

**Comment by poja** [comment]
No, didn't try

**Comment by barakugav** [comment]
[221121_180834.csv](https://github.com/poja/Cattus/files/10059127/221121_180834.csv)
[hex4_1121_1808.txt](https://github.com/poja/Cattus/files/10059128/hex4_1121_1808.txt)
Fixing the axis of batch norm seems to solve the problem
